### ISSUE TITLE
Improve Documentation for SoftSeg Training

### ISSUE THREAD

**Issue by naga-karthik** [ivadomed/ivadomed#1056]
### Motivation for the feature

The current documentation is not "complete" when it comes to the option of training with soft segmentations. Currently, the key "soft_gt" just says that setting it to _true_ is an indication that SoftSeg will be used for training and the GT masks are not binarized.    

### Description of the feature

While that is a good description, it does not say what other keys in the config file should be changed in order to have the full effect on the SoftSeg training. In particular, there are 2 things:

1. The key "[final_activation](https://ivadomed.org/configuration_file.html#soft-gt)" must be set to "relu", which uses normalized ReLU as the final layer. 
2. The postprocessing key "[binarize_prediction](https://ivadomed.org/configuration_file.html#binarize-prediction)" should be set to "-1". 

Note that the documentation for "binarize_prediction" says that it should be set to -1 to use soft predictions. However, it would be helpful if the above modifications to the config file are indicated right in the documentation for "soft_gt" since that's what is the most important key for training using SoftSeg (as discussed by me and @uzaymacar).

I can open a small PR that makes these changes if this would be useful for others.

**Comment by andreanne-lemay** [comment]
Thanks for opening this issue! Another important aspect of SoftSeg is the loss function. It should be a regression loss (e.g., AdapWing loss).

**Comment by naga-karthik** [comment]
Thanks for the suggestion, @andreanne-lemay! I'll remember to include this as well. 

Also, I trained a SoftSeg model with AdapWingLoss and I observed that the training and validation losses are on the order of 10,000s! Have you also experienced something like this?

**Comment by andreanne-lemay** [comment]
Yes it's normal. There is not a minimum / maximum limit with AdapWingLoss (unlike the Dice loss).

**Comment by uzaymacar** [comment]
I have been incorrectly using the term SoftSeg to refer to soft segmentation in general, but as @andreanne-lemay notes the term SoftSeg specifically refers to soft GTs + ReLU + AdapWingLoss as shown in the [original paper](https://arxiv.org/pdf/2011.09041.pdf).

That being said, both in [ivadomed/ms-challenge-2021](https://github.com/ivadomed/ms-challenge-2021) and [ivadomed/model_seg_ms_mp2rage](https://github.com/ivadomed/model_seg_ms_mp2rage) I observed the combination of soft GTs + ReLU + DiceLoss giving better results compared to soft GTs + ReLU + AdapWingLoss. In fact, converging with AdapWingLoss seemed like a challenge in these projects. 

I intend to debug AdapWingLoss / see where I'm doing wrong in the near-future. However, in the context of this issue I would like to point out that a user might abondon the soft segmentation approach if they observe a similar non-convergence with AdapWingLoss, whereas they could have observed performance gains if they used DiceLoss instead. I would therefore be in favor of updating documentation such that (a) the user is aware that they can use soft GTs + ReLU + DiceLoss for soft segmentation (in a loss-agnostic way) and (b) they are aware they can use AdapWingLoss instead of DiceLoss in case they want to implement SoftSeg as implemented in the original paper.

**Comment by naga-karthik** [comment]
@uzaymacar That sounds like a good plan! I agree about the user-abandonment issue, it's common to expect the loss to be between 0-1 or in the order of 10s but having large loss values might raise some concern over its "correctness" and hence the user might switch to DiceLoss instead. 

Just so we are on this topic, I am getting a strange warning on using the "relu" as the final activation layer:
```
TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  out = nn.ReLU()(seg_layer) / nn.ReLU()(seg_layer).max() if bool(nn.ReLU()(seg_layer).max()) \

```
I am wondering whether this is expected because a tensor is forcefully converted to a boolean. However, I do not recall having this issue while training for the [ivadomed/ms-challenge-2021](https://github.com/ivadomed/ms-challenge-2021). Any thoughts on this?

**Comment by uzaymacar** [comment]
> However, I do not recall having this issue while training for the ivadomed/ms-challenge-2021. Any thoughts on this?

I don't recall as well. The only difference I see in our [implementation](https://github.com/ivadomed/ms-challenge-2021/blob/3e8e5b292cbc64edcda7add17a11654969d1d080/modeling/models.py#L527) is that we called the ReLU function through `torch.nn.functional`. That might solve it, but if not, if I'm understanding correctly the `bool(...)` call functions to check if the maximum activation value is non-zero or not as zero would yield division by zero error. Therefore, perhaps changing that to something simpler like `nn.ReLU()(seg_layer).max().abs() > 0.0` might also solve the issue?

**Comment by naga-karthik** [comment]
Thanks for the suggestions, @uzaymacar! I tried them both and I still got the TracerWarning. An important note is that I get this error only after the training is done (not during the forward pass when the final activation is also calculated). This suggested that the error is not in how the forward pass output is calculated, rather, it has something to do with the export to the `onnx` format as also discussed in [this](https://github.com/onnx/onnx/issues/2836#:~:text=about%20this%20project%3F-,TracerWarning%3A%20Converting%20a%20tensor%20to%20a%20Python%20index%20might%20cause,not%20generalize%20to%20other%20inputs!) issue.

**Comment by uzaymacar** [comment]
Closing as #1064 was already approved and merged!
